### DISCUSSION TITLE
Add (optional) msRefreshBeforeExpires to SDK refresh token workflow again

### DISCUSSION THREAD

**Discussion by lubrst** [directus/directus#12636] [Feature Requests]
Hello everyone

In #12399 (Release 9.8.0) the msRefreshBeforeExpires parameter was removed and the SDK only refreshes the token when it has expired.

Here my thoughts on possible edge cases:

I use the token as access_token in an asset download request (not handled by the SDK, e.g. /assets/file_id?download&access_token=...) and call the new refreshIfExpired() function manually before this request.

What happens if the token is not expired in the refreshIfExpired() call, but expires just before the asset request?

```
await directus.auth.refreshIfExpired() 
// token expires now
downloadFile() // error
```

Can a optional msRefreshBeforeExpires paramter added again to ensure the refresh is done if the expiration time is close?

Current implementation:

https://github.com/directus/directus/blob/069e5eae1610cfbd6028f905dd94a860287e5a2d/packages/sdk/src/base/auth.ts#L71

Possible change:

`
if (this._storage.auth_expires_at < new Date().getTime() + msRefreshBeforeExpires)
`

 Default value of msRefreshBeforeExpires could be 0.

**Comment by lubrst** [comment]
A small addition with following scenario:

1. A page of my vuejs webapp is open until the access token expires.
2. Then a router link is clicked. Multiple simultaneous api requests are executed.
3. The refresh request is also triggered, but (some) of the other requests fail with 403.

It seems that requests are executed with an empty authorization header. I am not sure if this is an SDK bug.

In SDK versions prior to 9.8.0, this was not a problem, because the refresh was performed 30s before the token expired.

For this issue see #12648

**Comment by rijkvanzanten** [comment]
Thoughts @krazyjakee? 🙂

**Comment by krazyjakee** [comment]
For your first use case, this makes sense. Although it appears to be edge-case, it could certainly happen regularly at scale.
For your other issue, it's strange the auth header wouldn't be included at all? If it was introduced with the latest SDK changes then I suppose it makes sense that it could be related.

I propose that we re-introduce `msRefreshBeforeExpires` but I also propose the default be `30000` because I feel it's serious enough to require some kind of offset.

**Reply by lubrst** [reply]
The auth header is set, but the string is empty. 

```
beforeRequest: async (config) => {
	await this._auth.refreshIfExpired(); -> return immediately if storage reset is performed in refresh()
	const token = this.storage.auth_token; --> null if storage reset performed in refresh() and token not set yet
       ...
},
```

In this PR #12649 I have made a change in refreshIfExpired() to fix this issue.

**Reply by lubrst** [reply]
> For your first use case, this makes sense. Although it appears to be edge-case, it could certainly happen regularly at scale. For your other issue, it's strange the auth header wouldn't be included at all? If it was introduced with the latest SDK changes then I suppose it makes sense that it could be related.
> 
> I propose that we re-introduce `msRefreshBeforeExpires` but I also propose the default be `30000` because I feel it's serious enough to require some kind of offset.

I re-added `msRefreshBeforeExpires` in #12649
